### PR TITLE
Make SPIFFS and LittleFS stay out of link when not needed

### LINKED ISSUE #6691: LittleFS library linked when not used
I've been browsing through my builds using the very nice and new Inspect tool from the (not yet released) PIO 4.1.0rc1 which allows me to browse through the binary and see all libs included and its size.  (big, big kudo's for that tool!)

I noticed the LittleFS lib is linked even though I am not using it in my build (as far as I know)
Is this needed to be linked in when only using SPIFFS?
It does occupy about 20k in the build (compiled with debug, so release size my be less)

image

Apparently it is included here: https://github.com/esp8266/Arduino/blob/3733ece7e8c4ece1aaa99cef98dc9152bb277539/libraries/ESP8266HTTPUpdateServer/src/ESP8266HTTPUpdateServer-impl.h#L8

The same applies to ESP8266SdFat and SDFS.  (not 100% sure that's not included by accident)

### PR DESCRIPTION
**Pull request by d-a-v**
- define two weak functions defaulting to no-op
- redefine them to do something useful when either spiffs or littlefs are used/linked

This comes after a discussion with @earlephilhower 
(who will be too busy for the next days to come to continue on this)

replaces and closes #6697
fixes #6691 

Tested with all 4 uncommented `ets_printf` from this PR,
calling the two weak functions in `setup()`, and calling or not `SPIFFS.begin()` / `LittleFS.begin()`.

### CODE REVIEW

**Comment by TD-er** [comment]
I've looked at the code of this PR and I do think I get how it should work.
I'm not entirely sure about why it will not link the strong function implementations, but that's more lack of knowledge on my part I guess.

Do you already have any idea on how much smaller a SPIFFS-only (or LittleFS-only) build is with this fix?

**Comment by d-a-v** [comment]
SPIFFS is 32KB and LittleFS 25KB as far as I remember.

The trick in this PR and WPS is that there are `.cpp.o` files always linked in which we put the weakref functions. Other `.cpp.o` are indeed compiled but may not be included because their symbols are not needed (no dependencies). Once a dependency is forced in a linker unit (ie. call to `someInstance.begin()`, all of its symbols come to the final binary, and weak functions are overridden at that time if their new implementation are brought with.

**Comment by TD-er** [comment]
But why doesn't it try to link in the .o file if it does find the strong function implementation in there (and nothing else is needed)
Is it purely because the weak link already is resolved by something that is guaranteed to be already processed? How is that guaranteed?

**Comment by d-a-v** [comment]
> But why doesn't it try to link in the .o file if it does find the strong function implementation in there (and nothing else is needed)

I don't know that for sure.

> Is it purely because the weak link already is resolved by something that is guaranteed to be already processed? How is that guaranteed?

What I can say is that `close_all_fs()` needs to be linked, and it brings the weak definitions with it.

**Comment by devyte** [comment]
@TD-er it links either both, in which case the weak is discarded due to the presence of the strong, or just the weak, in which case the weak is used (and optimized out because it's an empty function).

**Comment by TD-er** [comment]
> @TD-er it links either both, in which case the weak is discarded due to the presence of the strong, or just the weak, in which case the weak is used (and optimized out because it's an empty function).

That's something I know, but what I don't know is how it is guaranteed the weak one is found first before the strong one, when the strong one is not needed.
The strong one will also be compiled, but the other functions in the .o file will not be needed when not called. So either linking the strong one or the weak one will resolve the link requirements, so I can imagine it depends on what order the function will be found by the linker.
Or am I missing something here?

**Comment by d-a-v** [comment]
When a strong function is present, the weak function is used when
- both weak and strong are in a library (`.a`)
- the weak function library is listed *before* the strong function library
- no other symbol from the strong object file is imported by linker

Example:
```
decls.h
===========

void func1 ();
void bring_strong_to_me();

main.cc
===========

#include "decls.h"

int main ()
{
#ifdef bringIt
    bring_strong_to_me();
#endif
    func1();
    
    return 0;
}
strong.cc
===========

#include <stdio.h>
#include "decls.h"

void bring_strong_to_me ()
{
}

void func1 ()
{
    printf("strong\n");
}
weak.cc
===========

#include <stdio.h>
#include "decls.h"

void func1 () __attribute__((weak));
void func1 ()
{
    printf("weak\n");
}
```

Run & results:
```
#!/bin/sh
set -e
CXX='g++ -Wall -Wextra -Werror'
for i in weak strong; do
    g++ -Wall -Wextra -Werror -c $i.cc -o $i.o
    ar rcs lib$i.a $i.o
done
set -x
$CXX           main.cc     strong.o weak.o -o func && ./func # -> strong
$CXX           main.cc     weak.o strong.o -o func && ./func # -> strong
$CXX           main.cc -L. -lstrong -lweak -o func && ./func # -> strong
$CXX           main.cc -L. -lweak -lstrong -o func && ./func # -> weak
$CXX -DbringIt main.cc -L. -lweak -lstrong -o func && ./func # -> strong
```

**Comment by TD-er** [comment]
Thanks.
That's a very clear example.

**Comment by d-a-v** [comment]
Have you seen a decrease in binary size with this PR ?
If not, maybe the linking order in PIO has to be examined.

**Comment by TD-er** [comment]
Have not yet tried this PR, since I have not been near a compile PC today.

**Comment by TD-er** [comment]
I did test it and it does save about 20k in build size (using SPIFFS)
Just updated one of my nodes and it works just fine.

File: cores/esp8266/spiffs_api.cpp [hunk 1]
```diff
@@ -127,6 +127,12 @@ bool isSpiffsFilenameValid(const char* name)
 
 }; // namespace
 
+extern "C" void spiffs_weak_end(void)
```
Comment (devyte) [inline cores/esp8266/spiffs_api.cpp]: Function name is a bit misleading: this function is not weak. 
I was going to suggest just using `spiffs_end()` and `littlefs_end()` across the board, but I find it does make sense to have the name reflect the weak symbol trick. 

What do you think of something like:
`spiffs_end_weaklink()`
?

File: libraries/LittleFS/src/LittleFS.cpp [hunk 2]
```diff
@@ -181,6 +181,12 @@ int LittleFSImpl::lfs_flash_sync(const struct lfs_config *c) {
 
 }; // namespace
 
+extern "C" void littlefs_weak_end(void)
+{
+    //ets_printf("debug: not weak littlefs end\n");
+    LittleFS.end();
```
Comment (devyte) [inline libraries/LittleFS/src/LittleFS.cpp]: what happens when FS_NO_GLOBALS is defined? NO GLOBAL_INSTANCES? etc
what happens when FS_NO_GLOBALS etc  is defined and there is a manual instance of a FS?

File: cores/esp8266/spiffs_api.cpp (hunk 1 above)
Comment (TD-er) [reply cores/esp8266/spiffs_api.cpp]: It is for internal call only right?
Then it does indeed need some reference to the weak trick.

File: libraries/LittleFS/src/LittleFS.cpp (hunk 2 above)
Comment (TD-er) [reply libraries/LittleFS/src/LittleFS.cpp]: I think it would make sense to have the call to `end()` placed in the destructor.

Comment (d-a-v) [reply libraries/LittleFS/src/LittleFS.cpp]: @devyte thanks. override is now only enabled with global instance is defined.
@TD-er littlefs already has a destructor, spiffs now calls `end()` on destructor

File: cores/esp8266/spiffs_api.cpp (hunk 1 above)
Comment (d-a-v) [reply cores/esp8266/spiffs_api.cpp]: Functions are renamed, indeed `weak` should not appear in their name.
A link to the snippet above is added in source code.

**Comment by devyte** [comment]
@TD-er please retest. If ok, this can be merged now for 2.6.0.

**Comment by TD-er** [comment]
> @TD-er please retest. If ok, this can be merged now for 2.6.0.

I will be back home Thursday evening so I cannot test it before.

**Comment by Jason2866** [comment]
We dont use SPIFFS nor LittleFS in Tasmota

**Comment by Jason2866** [comment]
A short test sketch with SPIFFS worked. Only SPIFFS linked
image

**Comment by Jason2866** [comment]
**Time to release 2.6.0 :-)** Tasmota does work reliable with latest master from tonight.
We have best experience using SDK22x_191024 with Tasmota.
With SDK22x_190703 there are a few devices (not many!) which have wifi reconnects
This devices do work too when using SDK22x_191024!

**Comment by d-a-v** [comment]
We are going to release very soon now, probably this week.
Thanks for the report on the Nonos SDK version.
I'm now wondering if default fw should be 1910 or 1907.

**Pull request merged**